### PR TITLE
Upgrade build_barback to build 0.9.0

### PR DESCRIPTION
**Pull request by natebosch**
- Change expectedOutputs to an Iterable. Some use cases need to check
  for duplicates. This means it will be illegal for the
  `buildExtensions` to express two ways to output the same file.
- Take `buildExtensions` as an argument to `TransformerBuilder`. We
  cannot automatically translate from declareOutputs to the new
  approach. This does mean that non DeclaringTransfomer instances can be
  wrapped.
- Update Builder, AssetReader, and AssetWriter implementations for the
  breaking changes in build

### CODE REVIEW

File: build_barback/lib/src/transformer/transformer.dart [hunk 1]
```diff
@@ -137,15 +138,13 @@ class _TransformAssetWriter implements AssetWriter {
   _TransformAssetWriter(this.transform);
 
   @override
-  Future writeAsBytes(build.AssetId id, List<int> bytes) {
-    transform.addOutput(barbackAssetFromBytes(id, bytes));
-    return new Future.value(null);
+  Future writeAsBytes(build.AssetId id, FutureOr<List<int>> bytes) async {
```
Comment (jakemac53) [inline build_barback/lib/src/transformer/transformer.dart]: I think it is important that this happens synchronously to make sure it gets waited on later?

You could verify by adding a test that returns synchronously from `build` after calling this method.

File: build_barback/lib/src/transformer/transformer.dart [hunk 2]
```diff
@@ -137,15 +138,13 @@ class _TransformAssetWriter implements AssetWriter {
...
+    transform.addOutput(barbackAssetFromBytes(id, await bytes));
   }
 
   @override
-  Future writeAsString(build.AssetId id, String contents,
-      {Encoding encoding: UTF8}) {
-    transform.addOutput(barbackAssetFromString(id, contents));
-    return new Future.value(null);
+  Future writeAsString(build.AssetId id, FutureOr<String> contents,
```
Comment (jakemac53) [inline build_barback/lib/src/transformer/transformer.dart]: ditto

File: build_barback/lib/src/transformer/transformer.dart (hunk 1 above)
Comment (natebosch) [inline build_barback/lib/src/transformer/transformer.dart]: The `.complete()` in `BuildStep` will wait for this so I don't think it matters here.

I'm also considering not changing `AssetWriter` and only changing the interface in `BuildStep`. That way most of our implementations won't need to care about it and there are only a few packages which see those directly anyway.

Comment (jakemac53) [inline build_barback/lib/src/transformer/transformer.dart]: Either way sounds good to me.

Comment (natebosch) [inline build_barback/lib/src/transformer/transformer.dart]: Reverted this part. AssetWriter doesn't take FutureOr anymore.

**Pull request merged**